source/common/stats/stats_impl.h

```diff
@@ -333,27 +339,62 @@ class GaugeImpl : public Gauge, public MetricImpl {
   // Stats::Gauge
   virtual void add(uint64_t amount) override {
     data_.value_ += amount;
-    data_.flags_ |= RawStatData::Flags::Used;
+    data_.flags_ |= Flags::Used;
   }
   virtual void dec() override { sub(1); }
   virtual void inc() override { add(1); }
   virtual void set(uint64_t value) override {
     data_.value_ = value;
-    data_.flags_ |= RawStatData::Flags::Used;
+    data_.flags_ |= Flags::Used;
   }
   virtual void sub(uint64_t amount) override {
     ASSERT(data_.value_ >= amount);
     ASSERT(used());
     data_.value_ -= amount;
   }
   virtual uint64_t value() const override { return data_.value_; }
-  bool used() const override { return data_.flags_ & RawStatData::Flags::Used; }
+  bool used() const override { return data_.flags_ & Flags::Used; }
 
 private:
   RawStatData& data_;
   RawStatDataAllocator& alloc_;
 };
 
+/**
+ * Implementation of HistogramStatistics for circllhist.
+ */
+class HistogramStatisticsImpl : public HistogramStatistics {
+public:
+  HistogramStatisticsImpl() : computed_quantiles_(supported_quantiles_.size(), 0.0) {}
+
+  HistogramStatisticsImpl(histogram_t* histogram_ptr)
+      : computed_quantiles_(supported_quantiles_.size(), 0.0) {
+    hist_approx_quantile(histogram_ptr, supported_quantiles_.data(), supported_quantiles_.size(),
+                         computed_quantiles_.data());
+  }
+
+  std::string summary() const override {
+    std::stringstream summary_stream;
+    size_t index = 0;
+    for (double quantile : supported_quantiles_) {
+      summary_stream << "P" << quantile * 100 << ": " << computed_quantiles_[index];
+      if (index < supported_quantiles_.size() - 1) {
+        summary_stream << ", ";
+      }
+      index++;
+    }
+    return summary_stream.str();
+  }
+
+  const std::vector<double>& supportedQuantiles() const override { return supported_quantiles_; }
+
+  const std::vector<double>& computedQuantiles() const override { return computed_quantiles_; }
+
+private:
+  std::vector<double> supported_quantiles_ = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```

Comment:
const

Reply:
i tried making it const earlier. but if I make it, I need to use const_cast to pass to hist_approx_quantile method and after that I get bunch of errors like this
copy assignment operator of 'HistogramStatisticsImpl' is implicitly deleted because field 'supported_quantiles_' has no copy assignment operator

Reply:
If hist_approx_quantile method doesn't declare its read-only args const, perpahs @postwait could address that upstream?

And just go ahead and explicitly delete the copy-ctor and operator= methods: https://stackoverflow.com/questions/6077143/disable-copy-constructor

Reply:
Addressed upstream... the input `double *` for quantiles requested is now a `const double *`

Reply:
@jmarantz if I delete assignment operator, I need to assign it after merge like this
cumulative_statistics_ = HistogramStatisticsImpl(cumulative_histogram_);
Is there any alternative to it?

Reply:
Yes: I'd declare it as:
```
   std::unique_ptr<HistogramStatisticsImpl> cumulative_statististics_;
```
and then assign it as
```
   cumulative_statistics_ = std::make_unique<HistogramStatisticsImpl>(cumulative_histogram_);
```
that way you won't be copying any histograms by value.

Reply:
In the last review cycle I changed this from a ptr to heap allocation as @mattklein123 felt it is better that way. I can change it back to pointer though.

Reply:
I haven't 'suggested any change to how the vector<double> gets allocated, just how the class containing it gets managed.

Reply:
sure. agree. not in this class, in the place where this is used, it was pointer. No issues. I will change to pointer again.

Reply:
I'm not sure what comment of @mattklein123's you are referring to; if there is a disagreement go with his approach as I'm still learning this codebase myself; I'm just always paranoid about copying classes by value without scrutinizing that it's an appropriate operation, and one thing to be particularly careful of is member variables to objects managed by C libraries.

WRT this particular member var, I'm curious if a simple syntax change would allow you to copy-construct it as a const, declare the value as:
```
 const std::vector<double> supported_quantiles_{0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```
I'm not sure if that will work or help as I'm actually new to some of the more modern constructs like initializing member variables where they are declared.  You could also get the copy-ctor to work by making an explicit one:
```
  HistogramStatisticsImpl(const HistogramStatisticsImpl& src)
    : supported_quantiles_(src.supported_quantiles_) {
  }
```
but you have to be diligent about auditing all the member vars and making sure it makes sense to copy them.  Move-semantics with std::unique_ptr on the class object is a lot easier.

Another model is to keep the HistogramStatisticsImpl allocated by value where its contained, as you have it now, and not re-assign it, but instead have a clear() method and re-populate it as needed.  That's actually the most efficient approach in that you don't have to copy that vector, though you may need to clear your accumulator values.

Reply:
I thought about this yesterday again. I ended up adding refresh method that takes the new histogram. It clears the existing one and recomputes again. So now there will be only Statistics object per histogram